### ISSUE TITLE
Entropy Regularization in PPOTrainer

### ISSUE THREAD

**Issue by DaehanKim** [lvwerra/trl#131]
[This youtube video](https://www.youtube.com/watch?v=1ppslywmIPs) suggests adding entropy regularization (simply adding average negative entropy * some coefficient to the loss) helps ppo agent explore better. Also Stable Baseline 3 also adopts [entropy regularization](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/ppo/ppo.py#L250).

Do you think this would work for LM too? If action spaces are too big, restricting action space to top-p or top-k would be helpful.

**Comment by natolambert** [comment]
Most of these things are really poorly documented in the RLHF space. If I clean up #119 , could we track all of them there with experiment logs on performance?

A bug fix in the control notebook is being integrated soon #126, so maybe evaluation will get more interesting.

**Comment by lvwerra** [comment]
Restricting top-p/k can help generating better sequences. We can also play with temperature to force more exotic generations. Beware that if you change the generation temperature you would also need to take this into account when computing the log-probs for the KL-divergence penalty (made that mistake once before 😄).

**Comment by DaehanKim** [comment]
Thank you guys for the specific comments!
@lvwerra By retricting action space to top-p or top-k, I meant giving entropy loss only considering those vocabs of interests!
i.e. we have 50k vocabs for softmax output. but letting the model explore with naive entropy loss through all those 50k vocabs might cause degeneration of texts by accidentally picking up a low-probability token. So Computing entropy with masked softmax output(that masks out logits not included in top-k or top-p) would help explore only in meaningful vocab spaces. 

Also since you mentioned, I was curious why all examples use simple categorical sampling for generation. As I see it, there are kinda standard decoding methods for open-ended generation and short text generations (with reference). These might be p-sampling (or momentum sampling, contrastive search if these can be integrated), and beam search or greedy search with repetition penalty. It would be beneficial to explore how these decoding methods affect RL.

**Comment by lvwerra** [comment]
Main reason I have become quite conservative with sampling during generation is that I got burnt a few times. The main issue is that if you don't sample from the natural distribution of tokens is that there is no guarantee that KL-div term is positive. Since you use it as a penalty the model can exploit negative KL-div (since it effectively becomes a positive reward).

One example where this happened: if you constrain the minimum sequence length the EOS token is excluded from sampling until you reach the minimum length. What happened is that the model learned to assign very high probabilities to that token making the probabilities for all other tokens (including the sampled ones) very low which led to a negative reward. I could imagine similar subtle scenarios for top-p/k sampling but it's worth exploring.

**Comment by DaehanKim** [comment]
Thank you for sharing your experience. @lvwerra
I think negative KL can be avoided by masking out and then renormalizing each distribution so that they become valid probability distribution again. For instance, `ppo_policy = [0.1,0.1,0.8]` and `ref_policy = [0.05, 0.15, 0.8]` and `word_2` is masked out (for ease of computation). `ppo_policy = [0.1, 0.1] -> [0.5, 0.5]` and `ref_policy = [0.25, 0.75]` can be renormalized to form valid probability distributions. Hopefully this scheme can work in top-k or top-p scenario. 
   
In case of min-length restriction, EOS logit in reference_distribution can also be masked out when comparing KL-div to prevent illegal inflation of EOS probability in ppo agent. (Though this needs subtle implementation and adds complication to the library for sure)

So what I'm thinking for top-p implementation( and anologously to top-k) is to generate responses with top-p scheme first, and logits are masked out with same top-p scheme before computing log probability. I can slightly change topPLogitWarper in transformer library to return logit masks that is applied to ppo agent logits. This same mask can be applied to ref_logits to generate same masked distribution and do log probabilty and kl-div computation. Do you think this is worth trying?

**Comment by DaehanKim** [comment]
I added entropy regularization in PPOTrainer and ran experiments with `sentiment-control.ipynb` script.
I didn't use any sampling scheme but pure categorical sampling.
Seems like it's not dramatic in this example 😁 but makes agent to explore more (when looking at the lm-head entropy).

![image](https://user-images.githubusercontent.com/20675681/218455616-375305b8-30f1-44de-9fda-1c4717e92412.png)


![image](https://user-images.githubusercontent.com/20675681/218455244-6f199bc4-2230-4da4-ac05-a9939268cd8c.png)

[full wandb chart here](https://wandb.ai/lucas01/trl?workspace=user-lucas01)

**Comment by lvwerra** [comment]
Nice set of experiments :) I guess it doesn't help much in that case but it's cool that you managed get it to work! Would you mind sharing the code change necessary? I don't think we want to integrate it into TRL at this point unless there is a sign of improvement but it would be useful to have the code somewhere in case someone wants to run another experiment.

**Comment by DaehanKim** [comment]
Sure it's a really simple change. 

First, you can add `entreg_coef` argument in PPOConfig initializer. [full code](https://github.com/DaehanKim/trl/blob/4f742f9d9a0f931600e5da8d0b5c886d5eb1a7c3/trl/trainer/ppo_config.py#L75-L99
)

```
@dataclass
class PPOConfig(object):
 def __init__(
        self,
        ...
       entreg_coef: Optional[float] = 0.0, # no entropy regularization by default
        ...
    ):
   ...
   self.entreg_coef = entreg_coef
   ...
```

Second, define your loss as follow : [full code](https://github.com/DaehanKim/trl/blob/4f742f9d9a0f931600e5da8d0b5c886d5eb1a7c3/trl/trainer/ppo_trainer_custom.py#L651-L652)

```
def train_minibatch(
        self,
        old_logprobs: torch.FloatTensor,
        values: torch.FloatTensor,
        rewards: torch.FloatTensor,
        query: torch.LongTensor,
        response: torch.LongTensor,
        model_input: torch.LongTensor,
    ):
        ...
        loss_p, loss_v, loss_ent, train_stats = self.loss(old_logprobs, values, rewards, logits, vpred, logprobs)
        loss = loss_p + loss_v + loss_ent
        ...
```
```
 def loss(
        self,
        old_logprobs: torch.FloatTensor,
        values: torch.FloatTensor,
        rewards: torch.FloatTensor,
        logits: torch.FloatTensor,
        vpred: torch.FloatTensor,
        logprob: torch.FloatTensor,
    ):
   ...
        # simply change the order of entropy computation to use it as a loss
        # and ent_loss is defined as negative entropy
        # it is multiplied with some coefficient and is added to the loss  
        entropy = torch.mean(entropy_from_logits(logits))
        ent_loss = - entropy # negative entropy to be minimised

        loss = pg_loss + self.config.vf_coef * vf_loss + self.config.entreg_coef * ent_loss

        approxkl = 0.5 * torch.mean((logprob - old_logprobs) ** 2)
        policykl = torch.mean(logprob - old_logprobs)
        return_mean, return_var = torch.mean(returns), torch.var(returns)
        value_mean, value_var = torch.mean(values), torch.var(values)

        # and ent_loss is added to the stat dict to plot it in wandb dashboard
        stats = dict(
            loss=dict(policy=pg_loss, value=vf_loss, ent_loss=ent_loss, total=loss),
            ...)
```

**Comment by lvwerra** [comment]
Thank you! Closing this issue for now, feel free to reopen if you find something new and interesting!

**Comment by leonardtang** [comment]
Can we add @DaehanKim's change as an official PR? Seems like a critical part of getting PPO to work?

**Comment by RylanSchaeffer** [comment]
@lvwerra @leonardtang I found this issue because my PPO runs constantly exhibit collapsing entropy:

- In general, is there a standard and generally successful solution to this problem?
- If so, is it already implemented in `TRL`?
